enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,612 @@
+that could brick the device, though it seems rather unlikely.
+Even if kernel panic occurs after staging a deployment with new configuration,
+thanks to greenboot functionality within the grub itself, the system will eventually rollback to
+previous deployment.
+Also, it is assumed that users are not pushing new image to production devices without prior verification on reference device.
+
+It may happen that some users need to use TuneD plugins that are not handled by the profile we'll create.
+In such case we may investigate if it's something generic enough to include, or we can instruct them
+to create new profile that would include `microshift-low-latency` profile.
+
+Systemd daemon we'll provide to enable TuneD profile should have a strict requirement before it
+reboots the node, so it doesn't put it into a boot loop.
+This pattern of reboot after booting affects the number of "effective" greenboot retries,
+so customers might need to account for that by increasing the number of retries.
+
+
+### Drawbacks
+
+Approach described in this enhancement does not provide much of the NTO's functionality
+due to the "static" nature of RPMs and packaged files (compared to NTO's dynamic templating),
+but it must be noted that NTO is going beyond low latency.
+
+One of the NTO's strengths is that it can create systemd units for runtime configuration
+(such as offlining CPUs, setting hugepages per NUMA node, clearing IRQ balance banned CPUs,
+setting RPS masks). Such dynamic actions are beyond capabilities of static files shipped via RPM.
+If such features are required by users, we could ship such systemd units and they could be no-op
+unless they're turned on in MicroShift's config. However, it is unknown to author of the enhancement
+if these are integral part of the low latency.
+
+## Open Questions [optional]
+
+- Verify if osbuild blueprint can override a file from RPM 
+  (variables.conf needs to exist for tuned profile, so it's nice to have some fallback)?
+- NTO runs tuned in non-daemon one shot mode using systemd unit.
+  Should we try doing the same or we want the tuned daemon to run continuously?
+- NTO's profile includes several other beside cpu-partitioning: 
+  [openshift-node](https://github.com/redhat-performance/tuned/blob/master/profiles/openshift-node/tuned.conf)
+  and [openshift](https://github.com/redhat-performance/tuned/blob/master/profiles/openshift/tuned.conf) - should we include them or incorporate their settings?
+- NTO took an approach to duplicate many of the setting from included profiles - should we do the same?
+  > Comment: Probably no need to do that. `cpu-partitioning` profile is not changed very often,
+  > so the risk of breakage is low, but if they change something, we should get that automatically, right?
+- Should we also provide NTO's systemd units for offlining CPUs, setting hugepages per NUMA node, clearing IRQ balance, setting RPS masks?
```

Comment:
I need to take a closer look at what these units are, but we should be able to do that with just TuneD as it's the original tool. Likely NTO had to create some workarounds for some functionality that wasn't possible in the traditional TuneD manner.

Reply:
For setting hugepages per NUMA the user can set a sysctl:
```
/sys/devices/system/node/node${NUMA_NODE}/hugepages/hugepages-${HUGEPAGES_SIZE}kB/nr_hugepages 
```

Similarly with off-lining CPUs:
```
# offiline
$ echo 0 | sudo tee /sys/devices/system/cpu/cpu${CPU_NUMBER}/online
```

Toggling CPUs offline and then back online is a relatively common thing that happens in low latency tuning to try force migrate any remaining IRQs off the isolated CPUs. Unfortunately the kubelet does not dynamically update which CPUs are online, so the CPUs that are online when the kubelet starts are the only ones that will be available for scheduling. 

Modifying the irqbalance CPUs must be done in a script as it's just doing file operations on `/etc/sysconfig/irqbalance`.

Reply:
I noticed that `cpu-partitioning` bans IRQ on CPUs equal to given `$isolated_cores`: https://github.com/redhat-performance/tuned/blob/master/profiles/cpu-partitioning/tuned.conf#L49-L50.
I think we can just reuse this, NTO just takes manual route.

I'll include offlining in the profile's script.

Reply:
> For setting hugepages per NUMA the user can set a sysctl...

Is this worth incorporating in this doc, or should/is this documented elsewhere (maybe worth linking)?

Reply:
I mentioned this specific thing. Offlining CPUs will be included in our profile. I'll link RHEL Tuned Documentation